src/meta-srv/src/metasrv.rs

```diff
@@ -176,17 +178,29 @@ pub struct MetaStateHandler {
     procedure_manager: ProcedureManagerRef,
     subscribe_manager: Option<SubscribeManagerRef>,
     greptimedb_telemetry_task: Arc<GreptimeDBTelemetryTask>,
+    leader_cached_kv_store: Arc<LeaderCachedKvStore>,
+    state: StateRef,
 }
 
 impl MetaStateHandler {
     pub async fn on_become_leader(&self) {
+        self.state.write().unwrap().next_state(become_leader(false));
+
+        if let Err(e) = self.leader_cached_kv_store.load().await {
+            error!(e; "Failed to load kv into leader cache kv store");
+        } else {
+            self.state.write().unwrap().next_state(become_leader(true));
+        }
```

Comment:
What if `self.state.write().unwrap().next_state(become_follower());` is called in other threads when `on_become_leader` is processing?  Lock `self.state` this way, I'm afraid there will be some data races in the future. I suggest use `self.state` to lock the entire function.

```rust
pub async fn on_become_leader(&self) {
    let state = self.state.write().unwrap();
    if let Err(e) = self.leader_cached_kv_store.load().await {
        error!(e; "Failed to load kv into leader cache kv store");
        state.next_state(become_leader(false));
    } else {
        state.next_state(become_leader(true));
    }
    // other logic...
}
```

This logic is also applied to the usage of `self.state` in `on_become_follower()`.

An addition question: `StateRef` is not using anywhere else in the code base, so what's the point of introducing this `struct State`?

Reply:
In this scenario, we only use a single thread.